iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -363,20 +362,20 @@ private Message ProcessMessage(bool hasMessage)
             finally
             {
                 if (Logging.IsEnabled)
-                    Logging.Exit(this, message, $"hasMessage={hasMessage}", nameof(ProcessMessage));
+                    Logging.Exit(this, message, $"Processed received C2D message with Id={message?.MessageId}", nameof(ProcessMessage));
             }
         }
 
-        private async Task<bool> ReceiveMessageArrivalAsync(CancellationToken cancellationToken)
+        private async Task WaitUntilC2dMessageArrivesAsync(CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
             CancellationToken disconnectToken = _disconnectAwaitersCancellationSource.Token;
             EnsureValidState();
 
             using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, disconnectToken);
 
-            // -1 millisecond represents for SemaphoreSlim to wait indefinitely until either of the linked cancellation tokens have been canceled.
-            return await _receivingSemaphore.WaitAsync(TimeSpan.FromMilliseconds(-1), linkedCts.Token).ConfigureAwait(true);
+            // Wait until either of the linked cancellation tokens have been canceled.
+            await _receivingSemaphore.WaitAsync(linkedCts.Token).ConfigureAwait(false);
```

Comment:
Does the behavior change here? If so, how? Or is this just a better way of waiting until token expires?

Reply:
The behavior doesn't change here, the semaphore still waits until it receives a message or either of the cancellation tokens are cancelled. I've explained the behavior in the PR description, let me know if it is unclear and I can update it.

Reply:
The OperationCancelledException is handled a bit differently based on where it is thrown from. For all protocols, if the OperationCancelledException is thrown as a result of cancellation of the passed cancellation token, it will get thrown back to the caller. For cancellation of the cancellation token generated from the default operation timeout (i.e. from the timeout API or no param API), the API will return a "null" message (this is because it wouldn't make sense to throw an OperationCancelledException from an API that doesn't accept a cancellation token.)

As far as the semaphore on MQTT is concerned, it doesn't really know if this is a user supplied cancellation token or the cancellation token generated from the default operation timeout; so it will simply throw the OperationCancelledException on cancellation, and delegate the handling of the exception to its caller.

Reply:
Got it, in that case the behavior of ReceiveAsync taking timeout as a parameter is changed correct?

Reply:
Synced offline, as per which the behavior of public API remains same.